File: csharp/Wjybxx.BigCat.Core/src/Fx/DefaultRpcProxyRegistry.cs
#region LICENSE

// Copyright 2025 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;

namespace Wjybxx.BigCat.Fx
{
/// <summary>
/// 默认的Rpc注册表实现
/// </summary>
public class DefaultRpcProxyRegistry : RpcProxyRegistry
{
    private readonly Dictionary<int, Delegate> proxyMap = new(512);
    private readonly Dictionary<int, RpcMethodInvoker> invokerMap = new(512);
    private readonly Dictionary<int, bool> disabledMap = new();
    private readonly Dictionary<int, object?> proxyDataMap = new(512);

    public void Register<T>(int serviceId, int methodId, RpcMethodProxy<T> proxy) {
        if (proxy == null) throw new ArgumentNullException(nameof(proxy));
        int methodKey = RpcMethodKey.MethodKey(serviceId, methodId);
        if (proxyMap.ContainsKey(methodKey)) {
            throw new ArgumentException($"methodKey is duplicate, serviceId: {serviceId}, methodId: {methodId}");
        }
        proxyMap[methodKey] = proxy;
        invokerMap[methodKey] = new RpcMethodInvoker<T>(proxy);
    }

    public virtual void SetProxyData(int serviceId, int methodId, object? customData) {
        int methodKey = RpcMethodKey.MethodKey(serviceId, methodId);
        if (customData == null) {
            proxyDataMap.Remove(methodKey);
        } else {
            proxyDataMap[methodKey] = customData;
        }
    }

    public object? GetProxyData(int serviceId, int methodId) {
        int methodKey = RpcMethodKey.MethodKey(serviceId, methodId);
        proxyDataMap.TryGetValue(methodKey, out object? proxyData);
        return proxyData;
    }

    public Delegate? GetProxy(int serviceId, int methodId) {
        int methodKey = RpcMethodKey.MethodKey(serviceId, methodId);
        proxyMap.TryGetValue(methodKey, out Delegate? proxy);
        return proxy;
    }

    public Delegate? RemoveProxy(int serviceId, int methodId) {
        int methodKey = RpcMethodKey.MethodKey(serviceId, methodId);
        proxyMap.Remove(methodKey, out Delegate proxy);
        invokerMap.Remove(methodKey, out _);
        return proxy;
    }

    public RpcMethodInvoker? GetInvoker(int serviceId, int methodId) {
        int methodKey = RpcMethodKey.MethodKey(serviceId, methodId);
        invokerMap.TryGetValue(methodKey, out RpcMethodInvoker? invoker);
        return invoker;
    }

    public void Disable(int serviceId, int methodId) {
        if (methodId == -1) {
            foreach (var pair in proxyMap) {
                if (RpcMethodKey.ServiceIdOfKey(pair.Key) == serviceId) {
                    disabledMap[pair.Key] = true;
                }
            }
        } else {
            int key = RpcMethodKey.MethodKey(serviceId, methodId);
            disabledMap[key] = true;
        }
    }

    public void Enable(int serviceId, int methodId) {
        if (methodId == -1) {
            List<int> keys = new List<int>();
            foreach (var pair in disabledMap) {
                if (RpcMethodKey.ServiceIdOfKey(pair.Key) == serviceId) {
                    keys.Add(pair.Key);
                }
            }
            foreach (int key in keys) {
                disabledMap.Remove(key);
            }
        } else {
            int key = RpcMethodKey.MethodKey(serviceId, methodId);
            disabledMap.Remove(key);
        }
    }

    public bool IsDisabled(int serviceId, int methodId) {
        int key = RpcMethodKey.MethodKey(serviceId, methodId);
        return disabledMap.ContainsKey(key);
    }

    public HashSet<int> Export() {
        return new HashSet<int>(proxyMap.Keys);
    }

    public void Clear() {
        proxyMap.Clear();
        invokerMap.Clear();
        proxyDataMap.Clear();
    }
}
}

File: java/framework/core/src/main/java/cn/wjybxx/bigcat/fx/DefaultRpcProxyRegistry.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.bigcat.fx;

import it.unimi.dsi.fastutil.ints.*;

import javax.annotation.Nonnull;
import java.util.Objects;

/**
 * 子类可重写{@link #setProxyData(int, int, Object)}提前解析数据
 *
 * @author wjybxx
 * date 2023/4/1
 */
public class DefaultRpcProxyRegistry implements RpcProxyRegistry {

    /**
     * 所有的Rpc请求处理函数, methodKey -> methodProxy
     */
    private final Int2ObjectMap<RpcMethodProxy<?>> proxyMap = new Int2ObjectOpenHashMap<>(512);
    private final Int2ObjectMap<Boolean> disabledMap = new Int2ObjectOpenHashMap<>();
    private final Int2ObjectMap<Object> proxyDataMap = new Int2ObjectOpenHashMap<>(512);

    @Override
    public <T> void register(int serviceId, int methodId, @Nonnull RpcMethodProxy<T> proxy) {
        Objects.requireNonNull(proxy, "proxy");
        final int methodKey = RpcMethodKey.methodKey(serviceId, methodId);
        if (proxyMap.containsKey(methodKey)) {
            throw new IllegalArgumentException("methodKey is duplicate, serviceId: %d, methodId: %d"
                    .formatted(serviceId, methodId));
        }
        proxyMap.put(methodKey, proxy);
    }

    @Override
    public void setProxyData(int serviceId, int methodId, Object customData) {
        final int methodKey = RpcMethodKey.methodKey(serviceId, methodId);
        if (customData == null) {
            proxyDataMap.remove(methodKey);
        } else {
            proxyDataMap.put(methodKey, customData);
        }
    }

    @Override
    public Object getProxyData(int serviceId, int methodId) {
        final int methodKey = RpcMethodKey.methodKey(serviceId, methodId);
        return proxyDataMap.get(methodKey);
    }

    @Override
    public RpcMethodProxy<?> getProxy(int serviceId, int methodId) {
        final int methodKey = RpcMethodKey.methodKey(serviceId, methodId);
        return proxyMap.get(methodKey);
    }

    @Override
    public RpcMethodProxy<?> removeProxy(int serviceId, int methodId) {
        final int methodKey = RpcMethodKey.methodKey(serviceId, methodId);
        return proxyMap.remove(methodKey);
    }

    @Override
    public void disable(int serviceId, int methodId) {
        if (methodId == -1) {
            for (var pair : proxyMap.int2ObjectEntrySet()) {
                if (RpcMethodKey.serviceIdOfKey(pair.getIntKey()) == serviceId) {
                    disabledMap.put(pair.getIntKey(), Boolean.TRUE);
                }
            }
        } else {
            int key = RpcMethodKey.methodKey(serviceId, methodId);
            disabledMap.put(key, Boolean.TRUE);
        }
    }

    @Override
    public void enable(int serviceId, int methodId) {
        if (methodId == -1) {
            IntList keys = new IntArrayList();
            for (var pair : disabledMap.int2ObjectEntrySet()) {
                if (RpcMethodKey.serviceIdOfKey(pair.getIntKey()) == serviceId) {
                    keys.add(pair.getIntKey());
                }
            }
            keys.forEach(disabledMap::remove);
        } else {
            int key = RpcMethodKey.methodKey(serviceId, methodId);
            disabledMap.remove(key);
        }
    }

    @Override
    public boolean isDisabled(int serviceId, int methodId) {
        int key = RpcMethodKey.methodKey(serviceId, methodId);
        return disabledMap.containsKey(key);
    }

    @Override
    public IntSet export() {
        IntOpenHashSet result = new IntOpenHashSet(10);
        proxyMap.keySet().intStream()
                .map(RpcMethodKey::serviceIdOfKey)
                .forEach(result::add);
        return result;
    }

    public void clear() {
        proxyMap.clear();
        proxyDataMap.clear();
    }

}